File: examples/java/patterns/saga-pattern/pitfall1-non-durable.java
// WRONG
for (int i = compensations.size() - 1; i >= 0; i--) {
    compensations.get(i).fn.run(); // if this throws error, subsequent compensations don't run
}

// CORRECT
for (int i = compensations.size() - 1; i >= 0; i--) {
    Compensation comp = compensations.get(i);
    try {
        context.step(comp.name, Void.class, ctx -> { comp.fn.run(); return null; });
    } catch (Exception e) {
        context.getLogger().error("Compensation failed: " + comp.name);
    }
}

File: examples/typescript/patterns/saga-pattern/pitfall1-non-durable.ts
// WRONG - not durable, not retried on failure
for (const comp of compensations.reverse()) {
  await comp.fn(); // if this throws error, subsequent compensations don't run
}

// CORRECT - each compensation is a durable step
for (const comp of compensations.reverse()) {
  try {
    await context.step(comp.name, async () => comp.fn());
  } catch (compError) {
    context.logger.error(`Compensation failed: ${comp.name}`, compError);
  }
}